iothub_client/src/iothubtransport_amqp_twin_messenger.c

```diff
@@ -1864,6 +1906,64 @@ int twin_messenger_report_state_async(TWIN_MESSENGER_HANDLE twin_msgr_handle, CO
     return result;
 }
 
+int twin_messenger_get_twin_async(TWIN_MESSENGER_HANDLE twin_msgr_handle, TWIN_STATE_UPDATE_CALLBACK on_get_twin_completed_callback, void* context)
+{
+    int result;
+
+    if (twin_msgr_handle == NULL || on_get_twin_completed_callback == NULL)
+    {
+        LogError("Invalid argument (twin_msgr_handle=%p, on_get_twin_completed_callback=%p)", twin_msgr_handle, on_get_twin_completed_callback);
+        // Codes_SRS_IOTHUBTRANSPORT_AMQP_TWIN_MESSENGER_09_109: [If `twin_msgr_handle` or `on_twin_state_update_callback` are NULL, twin_messenger_get_twin_async() shall fail and return a non-zero value]  
+        result = __FAILURE__;
+    }
+    else
+    {
+        TWIN_MESSENGER_INSTANCE* twin_msgr = (TWIN_MESSENGER_INSTANCE*)twin_msgr_handle;
+        TWIN_OPERATION_CONTEXT* twin_op_ctx;
+
+        if ((twin_op_ctx = create_twin_operation_context(twin_msgr, TWIN_OPERATION_TYPE_GET_ON_DEMAND)) == NULL)
+        {
+            LogError("Failed creating a context for TWIN request (%s, TWIN_OPERATION_TYPE_GET_ON_DEMAND)", twin_msgr->device_id);
+            // Codes_SRS_IOTHUBTRANSPORT_AMQP_TWIN_MESSENGER_09_113: [If any failures occurr, twin_messenger_get_twin_async() shall return a non-zero value ]  
+            result = __FAILURE__;
+        }
+        else
+        {
+            // Codes_SRS_IOTHUBTRANSPORT_AMQP_TWIN_MESSENGER_09_110: [ `on_get_twin_completed_callback` and `context` shall be saved ] 
+            twin_op_ctx->cb.get_twin.callback = on_get_twin_completed_callback;
+            twin_op_ctx->cb.get_twin.context = context;
+
+            if (add_twin_operation_context_to_queue(twin_op_ctx) != RESULT_OK)
+            {
+                LogError("Failed queueing TWIN request context (%s, TWIN_OPERATION_TYPE_GET_ON_DEMAND)", twin_msgr->device_id);
+
+                destroy_twin_operation_context(twin_op_ctx);
+                // Codes_SRS_IOTHUBTRANSPORT_AMQP_TWIN_MESSENGER_09_113: [If any failures occurr, twin_messenger_get_twin_async() shall return a non-zero value ]  
+                result = __FAILURE__;
+            }
+            // Codes_SRS_IOTHUBTRANSPORT_AMQP_TWIN_MESSENGER_09_111: [ An AMQP message shall be created to request a GET twin ]
+            // Codes_SRS_IOTHUBTRANSPORT_AMQP_TWIN_MESSENGER_09_112: [ The AMQP message shall be sent to the twin send link ] 
+            else if (send_twin_operation_request(twin_msgr, twin_op_ctx, NULL) != RESULT_OK)
+            {
+                LogError("Failed sending TWIN request (%s, TWIN_OPERATION_TYPE_GET_ON_DEMAND)", twin_msgr->device_id);
+
+                (void)remove_twin_operation_context_from_queue(twin_op_ctx);
+                destroy_twin_operation_context(twin_op_ctx);
```

Comment:
>twin_op_ctx [](start = 47, length = 11)

What destroys the twin operation context object outside this method? So far, I feel there might be leaks because there isn't a destructor for all the stuff associated to this context now, other than this amqp_twin_messenger logic. Could you explain / fix how the twin_op_ctx callback context is cleared after success, for example?

Reply:
1. on_amqp_send_complete_callback if result != AMQP_MESSENGER_SEND_RESULT_SUCCESS;
2. on_amqp_message_received_callback (search for destroy_twin_operation_context() )
3. remove_expired_twin_operation_request
4. cancel_all_pending_twin_operations

Please look at these to see where the context is destroyed.